### PR TITLE
MAT-3912: Save CQl definition for measure group populations

### PR DESCRIPTION
**Pull request by adongare**
## MADiE PR

Jira Ticket: [MAT-3912](https://jira.cms.gov/browse/MAT-3912)
(Optional) Related Tickets:

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [x] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [x] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [x] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/cms/gov/madie/measure/services/MeasureService.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  public Group createOrUpdateGroup(Group group, String measureId, String username) {
+    Measure measure = measureRepository.findById(measureId).orElse(null);
+    if (measure == null) {
+      throw new ResourceNotFoundException("Measure", measureId);
+    }
+    // no group present, this is the first group
+    if (CollectionUtils.isEmpty(measure.getGroups())) {
+      group.setId(ObjectId.get().toString());
```
Comment (sb-cecilialiu) [inline src/main/java/cms/gov/madie/measure/services/MeasureService.java:30]: Just curious, is this needed for add?

File: src/main/java/cms/gov/madie/measure/services/MeasureService.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    } else {
+      Optional<Group> existingGroupOpt =
+          measure.getGroups().stream().filter(g -> g.getId().equals(group.getId())).findFirst();
+      // if group already exists, just update it
+      if (existingGroupOpt.isPresent()) {
+        Group existingGroup = existingGroupOpt.get();
+        existingGroup.setScoring(group.getScoring());
+        existingGroup.setPopulation(group.getPopulation());
+      } else { // if not present, add into groups collection
+        group.setId(ObjectId.get().toString());
```
Comment (sb-cecilialiu) [inline src/main/java/cms/gov/madie/measure/services/MeasureService.java:41]: Same as above

File: src/main/java/cms/gov/madie/measure/services/MeasureService.java (hunk 1 above)
Comment (adongare) [reply src/main/java/cms/gov/madie/measure/services/MeasureService.java:30]: I think so. spring mongo data doesn't create an Id for Embedded documents. Assigning id manually helps to identify the subdocument by id during updates.

**Review by sb-cecilialiu — APPROVED** [review]
Looks great! Approving!

**Review by gregory-akins — APPROVED** [review]

File: src/main/java/cms/gov/madie/measure/models/MeasurePopulation.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package cms.gov.madie.measure.models;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public enum MeasurePopulation {
+  @JsonProperty("initialPopulation")
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/models/MeasurePopulation.java:6]: 👍

**Review by jkotanchik-SB — APPROVED** [review]

**Pull request merged**